### PR TITLE
[orc8r][tracing] Context propagation in all orc8r client libraries except configurator

### PR DESCRIPTION
**Pull request by xjtian**
## Summary

- Make client_api.go files context-aware and propagate contexts to GRPC calls
- No configurator yet because that one will require its own PR
- Mostly automated refactoring. bootstrapper servicer and identity decorator are the only interesting files in this diff

## Test Plan

- CI, `make test` locally

## Additional Information

- [ ] This change is backwards-breaking

Signed-off-by: Jacky Tian <jacky@xjtian.com>

### CODE REVIEW

**Review by themarwhal — APPROVED** [review]

File: fbinternal/cloud/go/services/testcontroller/statemachines/enodebd_test.go [hunk 1]
```diff
@@ -110,7 +110,7 @@ func Test_EnodebdE2ETestStateMachine_HappyPath(t *testing.T) {
 	// Check upgrade status, gateway hasn't upgraded yet
 	// ---
 	gatewayRecord := &models2.GatewayDevice{HardwareID: "hw1", Key: &models2.ChallengeKey{KeyType: "ECHO"}}
-	err = device.RegisterDevice("n1", orc8r.AccessGatewayRecordType, "hw1", gatewayRecord, serdes.Device)
+	err = device.RegisterDevice(context.Background(), "n1", orc8r.AccessGatewayRecordType, "hw1", gatewayRecord, serdes.Device)
```
Comment (hcgatewood) [inline fbinternal/cloud/go/services/testcontroller/statemachines/enodebd_test.go:113]: For all test files: in my mind, the first-best solution for tests is to have one context.Background shared across each test. That way, hypothetically, we could hook up jaeger with any particular misbehaving test to peer in better understand the issues

That said, if it's non-trivial effort to switch to this pattern, fk it and we can keep things simple

File: orc8r/cloud/go/services/certifier/client_api_test.go [hunk 2]
```diff
@@ -14,13 +14,15 @@ limitations under the License.
...
 import (
+	"context"
 	"crypto/x509"
 	"fmt"
 	"testing"
 	"time"
 
 	"github.com/golang/protobuf/proto"
 	"github.com/stretchr/testify/assert"
+	context2 "golang.org/x/net/context"
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/certifier/client_api_test.go:25]: For a followup PR: can you grep-replace all usages of this pkg, in favor of the stdlib `context`

File: lte/cloud/go/services/lte/obsidian/handlers/handlers_test.go [hunk 3]
```diff
@@ -1006,14 +1006,10 @@ func TestCreateGateway(t *testing.T) {
...
-		&models.GatewayDevice{
-			HardwareID: "hw2",
-			Key:        &models.ChallengeKey{KeyType: "ECHO"},
-		},
-		serdes.Device,
-	)
+	err = device.RegisterDevice(context.Background(), "n1", orc8r.AccessGatewayRecordType, "hw2", &models.GatewayDevice{
+		HardwareID: "hw2",
+		Key:        &models.ChallengeKey{KeyType: "ECHO"},
+	}, serdes.Device)
```
Comment (hcgatewood) [inline lte/cloud/go/services/lte/obsidian/handlers/handlers_test.go]: Here and elsewhere: ideally let's preserve/make uniform the formatting of these

Or, basically the min of what I'm looking for is newline after `(` and newline before `)`, other than that I care less

File: orc8r/cloud/go/services/accessd/accessd_test.go [hunk 4]
```diff
@@ -27,6 +28,7 @@ import (
 
 func TestAccessManager(t *testing.T) {
 	accessd_test_service.StartTestService(t)
+	ctx := context.Background()
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/accessd/accessd_test.go:31]: From earlier comment: yeah this was the test-based pattern I had in mind

File: orc8r/cloud/go/services/bootstrapper/servicers/bootstrapper.go [hunk 5]
```diff
@@ -282,6 +284,16 @@ func getChallengeKey(hwID string) (protos.ChallengeKey_KeyType, []byte, error) {
 	return protos.ChallengeKey_KeyType(keyType), key, nil
 }
 
+// Bootstrapper needs to ignore incoming SN/CN headers in ctx when it fans
+// out calls to upstream services, or else the identity middleware will trip.
+func strippedIncomingCtx(ctx context.Context) context.Context {
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/bootstrapper/servicers/bootstrapper.go:289]: 👍

File: orc8r/cloud/go/services/dispatcher/servicers/sync_rpc_service.go [hunk 6]
```diff
@@ -169,7 +169,7 @@ func (srv *SyncRPCService) receiveFromStream(
 			return
 		} else {
 			glog.V(2).Infof("HWID %v: processing response", coordinator.GwID)
-			err := srv.processSyncRPCResp(syncRPCResp, coordinator.GwID)
+			err := srv.processSyncRPCResp(coordinator.Ctx, syncRPCResp, coordinator.GwID)
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/dispatcher/servicers/sync_rpc_service.go:172]: Idk if this is the correct ctx but assuming you figured it out

File: orc8r/cloud/go/services/orchestrator/obsidian/handlers/gateway_handler_factory.go [hunk 7]
```diff
@@ -155,13 +155,14 @@ func GetReadGatewayDeviceHandler(path string, serdes serde.Registry) obsidian.Ha
 				return nerr
 			}
 
+			reqCtx := c.Request().Context()
```
Comment (hcgatewood) [inline orc8r/cloud/go/services/orchestrator/obsidian/handlers/gateway_handler_factory.go:158]: We're a little non-uniform with how we're naming/handling the `echo.Context` contexts. Not end of the world, but my preference would be: at the start of every echo handler func that needs to propagate its context, go ahead and grab `ctx := c.Request().Context()` at line 1 of the function, then pass that ctx in everywhere it's required

**Review by hcgatewood — CHANGES_REQUESTED** [review]
Thanks for the fix! Couple small nits. I'm down to stamp as-is, but preference would be to add the small couple cleanups here and going forward, for uniformity purposes. Lmk what you think

File: fbinternal/cloud/go/services/testcontroller/statemachines/enodebd_test.go (hunk 1 above)
Comment (xjtian) [reply fbinternal/cloud/go/services/testcontroller/statemachines/enodebd_test.go:113]: It's a lot of files being changed, so where you see context.Background() it's from an automated refactor to update the function signature (except in some edge cases). I think it'll be better to update these as needed the future once you have a concrete idea of how you'll want to configure tracing in unit tests.

File: lte/cloud/go/services/lte/obsidian/handlers/handlers_test.go (hunk 3 above)
Comment (xjtian) [reply lte/cloud/go/services/lte/obsidian/handlers/handlers_test.go]: yeah this is the refactorer's doing, I can do a once-over and check

File: orc8r/cloud/go/services/orchestrator/obsidian/handlers/gateway_handler_factory.go (hunk 7 above)
Comment (xjtian) [reply orc8r/cloud/go/services/orchestrator/obsidian/handlers/gateway_handler_factory.go:158]: I save the `reqCtx` variable when there's more than 1 place in the callsite that will need to propagate the context, otherwise I just use this inline at the client call

edit: I declare the variable as late as possible so as not to clutter the head of the functions with something that won't be used if the req fails validation for e.g.

**Comment by xjtian** [comment]
Updated PR to include all orc8r module services

File: orc8r/cloud/go/services/orchestrator/obsidian/handlers/gateway_handler_factory.go (hunk 7 above)
Comment (hcgatewood) [reply orc8r/cloud/go/services/orchestrator/obsidian/handlers/gateway_handler_factory.go:158]: In my mind, the context would always be the first defn -- to such a degree that ideally it wouldn't even need to be declared / would have language-level support for implicitly passing ctxs around

But in practice, I'm fine with this approach as well

File: feg/cloud/go/services/basic_acct/tests/setup.go [hunk 8]
```diff
@@ -146,11 +147,7 @@ func SetupNetworks(t *testing.T) {
 		serdes.Entity,
 	)
 	assert.NoError(t, err)
-	err = device.RegisterDevice(
-		FederatedLteNetworkID, orc8r.AccessGatewayRecordType, AgwHwId,
-		&models.GatewayDevice{HardwareID: AgwHwId, Key: &models.ChallengeKey{KeyType: "ECHO"}},
-		serdes.Device,
-	)
+	err = device.RegisterDevice(context.Background(), FederatedLteNetworkID, orc8r.AccessGatewayRecordType, AgwHwId, &models.GatewayDevice{HardwareID: AgwHwId, Key: &models.ChallengeKey{KeyType: "ECHO"}}, serdes.Device)
```
Comment (hcgatewood) [inline feg/cloud/go/services/basic_acct/tests/setup.go:150]: Here and elsewhere, for now or later: line too long, prefer to break up these types of lines in a canonical way

File: lte/cloud/go/services/lte/analytics/calculations/state.go [hunk 9]
```diff
@@ -129,6 +133,8 @@ func (x *SiteMetricsCalculation) Calculate(prometheusClient query_api.Prometheus
 	gatewayVersionCfg, gatewayVersionCfgOk := x.AnalyticsConfig.Metrics[metrics.GatewayMagmaVersionMetric]
 	enbConnectedCfg, enbConnectedOk := x.AnalyticsConfig.Metrics[metrics.EnodebConnectedMetric]
 
+	// TODO: you'll want to set some tracing params on this ctx
+	outgoingContext := context.TODO()
```
Comment (hcgatewood) [inline lte/cloud/go/services/lte/analytics/calculations/state.go:137]: For my understanding: why TODO here

File: orc8r/cloud/go/tools/accessc/handlers/add_existing.go [hunk 10]
```diff
@@ -90,7 +93,7 @@ func addExisting(cmd *commands.Command, args []string) int {
 				log.Fatalf(
 					"Cannot parse certificate from '%s': %s\n", certFName, err)
 			}
-			err = certifier.AddCertificate(operator, certDERBlock.Bytes)
+			err = certifier.AddCertificate(context.Background(), operator, certDERBlock.Bytes)
```
Comment (hcgatewood) [inline orc8r/cloud/go/tools/accessc/handlers/add_existing.go]: For all the accessc handlers: I'd imagined that each cmd should use a single ctx for the entirety of the cmd

**Review by hcgatewood — APPROVED** [review]
Lgtm, thanks for adding Jacky 🙂 

Main comment at this point is the per-CLI-handler contexts. But generally lgtm. Either way, msg me when you're ready and I'll merge for you

File: lte/cloud/go/services/lte/analytics/calculations/state.go (hunk 9 above)
Comment (xjtian) [reply lte/cloud/go/services/lte/analytics/calculations/state.go:137]: This is the point of origination for the network call, so you'd want to start your trace here and use that context. But that'll depend on your tracing client library

**Comment by xjtian** [comment]
@hcgatewood good to merge whenever
